### ISSUE TITLE
uniform block not compatible

### ISSUE THREAD

**Issue by gregory38** [apitrace/apitrace#75]
Hello,

It would be nice if apitrace could support Uniform Buffer Object. It allows to use buffer facilities instead of using single glUniform\* calls. See the extension for details : http://www.opengl.org/registry/specs/ARB/uniform_buffer_object.txt

1/ Uniforms are handled by block so you can't use anymore glGetUniformLocation/glGetUniform single call. From the spec:
i) An active uniform that is a member of a named uniform block has location == -1.
Here a patch to avoid to read random/uninitialized  value for uniform.
--------------------------->8----------------------------------
--- a/retrace/glstate_shaders.cpp
+++ b/retrace/glstate_shaders.cpp
@@ -220,9 +220,14 @@ dumpUniform(JSONWriter &json, GLint program, GLint size, GLenum type, const GLch

```
     std::string elemName = ss.str();
```
-        GLint location = glGetUniformLocation(program, elemName.c_str());
  +
-               if (location == -1) {
-                       continue;
- ```
            }
  ```
  
  +
       json.beginMember(elemName);
- ```
     GLint location = glGetUniformLocation(program, elemName.c_str());
  
   if (numElems > 1) {
       json.beginArray();
  ```
  
  ---------------------------8<----------------------------------
  2/ Full support of UBO which is only available on opengl 3.1+

A nice wiki that explain how to access uniform block info: http://www.opengl.org/wiki/GLSL_Uniforms#Accessing_uniform_information

In short: 
- glGetProgramiv was extended to support GL_ACTIVE_UNIFORM_BLOCKS and GL_ACTIVE_UNIFORM_BLOCK_MAX_NAME_LENGTH
- glGetActiveUniform was superseeded by glGetActiveUniformName and glGetActiveUniformsiv, the latter allow to get the "format" of the uniform inside the Uniform Block and the block index. So it would be possible to map the buffer, read it, then extra the uniform value with OFFSET, STRIDE ...
- It is possible to get the block name with the 2 commands: glGetActiveUniformBlockName and glGetActiveUniformBlockiv. It would be nice to have the block name inside qapitrace (with folding capability) but it might need some json format extensions.
